imsim/diffraction_fft.py

```diff
@@ -0,0 +1,226 @@
+
+
+def prepare_psf_field_rotation(
+    w: int,
+    h: int,
+    resolution: float,
+    alpha: float,
+    d_alpha: float,
+) -> np.ndarray:
+    """Spike PSF for finite angular spike with (field rotation).
+    w: Pixel width of the image
+    h: Pixel height of the image
+    resolution: length per pixel
+    alpha: Rotation angle of the spikes [rad]
+    d_alpha: angular spike width (field rotation angle)
+
+    Will return an image of the spike PSF of dimension (2w+1)x(2h+1).
+    The angular range of the spike will be [alpha, alpha + d_alpha] (4-fold symmetry).
+    """
+    x_range = np.arange(-w, w + 1)
+    y_range = np.arange(-h, h + 1)
+    [x, y] = xy = np.c_[
+        np.meshgrid(
+            x_range,
+            y_range,
+            indexing="ij",
+            copy=False,
+        )
+    ]
+    # Antialiased cross at alpha - d_alpha / 2:
+    psf = antialiased_cross(xy, alpha - d_alpha / 2.0)
+    # Interior:
+    th = np.arctan2(y, x)
+    dth = th - (alpha - d_alpha)  # angular distance to spike boundary
+    dth %= np.pi / 2  # 4-fold symmetry
+    psf[dth <= d_alpha] = 1.0
+    # Radial profile:
+    r = np.hypot(x, y) * resolution
+    psf *= field_rotation_profile(r, d_alpha, resolution)
+    # Normalize, so pixel values add up to 1.
+    # Here we actually are neglecting the non-visible tails of the cross.
+    psf /= np.sum(psf)
+    return psf
+
+
+def apply_diffraction_psf(
+    image: np.ndarray,
+    rottelpos: float,
+    exptime: np.ndarray,
+    latitude: float,
+    azimuth: float,
+    altitude: float,
+    full_well: float,
+    pixel_length: float,
+    psf_size: int,
+):
+    """2d convolve, leaving image dimension invariant.
+    image: Brightness pixel data
+    rottelpos: Telescope rotation angle [rad]
+    exptime: Exposure time (used to determine the spike width)
+    latitude: Geographic latitude of observators
+    azimuth: Azimuth of the telescope
+    altitude: Altitude of the telescope
+    full_well: Threshold value for saturated pixels
+    pixel_length: Scaling factor: The length of 1px in `spike_profile`
+    psf_size: Size of the PSF (width and height) which will be layed over saturated pixels.
+    """
+
+    psf_w = psf_h = psf_size  # neglect the outer region (~1/r^2)
+    sin_cos = np.empty(2)
+    e_z_0, e_z = prepare_e_z(latitude)
+    e_focal = e_equatorial(latitude=latitude, azimuth=azimuth, altitude=altitude)
+    field_rotation_sin_cos(e_z_0, e_z, e_focal, exptime, sin_cos)
+    d_alpha = np.arctan2(sin_cos[1], sin_cos[0])
+    rottelpos = np.pi/4. - rottelpos
+    psf = prepare_psf_field_rotation(
+        psf_w,
+        psf_h,
+        resolution=pixel_length,
+        alpha=rottelpos,
+        d_alpha=d_alpha,
+    )
+    for region_row, region_col in saturated_clusters(image, full_well):
+        img_region = image[region_row, region_col]
+        diffracted = scipy.signal.convolve2d(psf, img_region, mode="full")
```

Comment:
I think this part of the code would be significantly simpler if `psf` was set up to be the same size/shape as `image`.  Then you could call `convolve2d` with `mode='same'`, and the output would also be the same size as image.  Then adding is just `image += diffracted` without all the extra cropping.

Reply:
Regarding this and your next comment: Are you suggesting to just do a full convolution `image` * `pixel_spikes`, where `pixel_spikes` has the same dimension as `image`?

Reply:
No.  Still find a small region around the center of the PSF with all the >threshold pixels.  
But the thing you convolve with that could be the same size as the image or at least the part of the image that you do the convolution on (it could be some smaller image[bounds] if you want).  
The point is to let scipy handle all the edge details for you with the mode='same' option.

Reply:
Still not sure if I get it.
If I only take a small region with all the >threshold pixels, and make the cutoff spike psf that same size as the region, then the resulting spike width would be determined by the small region - i.e. in most cases << 4000px (="spike_width_cuttoff").
Or do you suggest to have the small region to be at least "spike_width_cuttoff"?

Reply:
OK. But then the resulting convolution still has to be put back to some region in the total image. This still would require cropping, or am I wrong?

Reply:
Sure, but that's much easier than what you are currently doing, since you can use the galsim.Bounds functionality.
```
b_spike = galsim.Bounds(star_pos).withBorder(spike_length_cutoff)
b_overlap = b_spike & image.bounds
spike_image = galsim.Image(spike_array, b_spike)
image[b_overlap] += spike_image[b_overlap]
```